### PR TITLE
ProductFormType should be extensible

### PR DESCRIPTION
**Pull request by boris-brtan**
| Q             | A
| ------------- | ---
|Description, reason for the PR| we want to have extensible product edit form via formType  classes
|New feature| Yes 
|BC breaks| Yes 
|Fixes issues| 
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Resources/views/Admin/Form/theme.html.twig [hunk 1]
```diff
@@ -588,7 +598,7 @@
                 '<li class="form-tree__item js-category-tree-form-item">' ~
                 form_widget(form.vars.prototype) ~
                 '<ul class="js-category-tree-form-children-container">' ~
-                '</ul></li>'|e
+                '</ul></li>'
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Resources/views/Admin/Form/theme.html.twig]: `/admin/product/new/`
image

File: packages/framework/src/Form/DisplayOnlyUrlType.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+                ],
+            ]);
+    }
+
+    /**
+     * @return null|string
+     */
+    public function getBlockPrefix()
+    {
+        return 'display_only_url';
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/DisplayOnlyUrlType.php]: yes, it works without blockPrefix

File: packages/framework/src/Form/LocalizedFullWidthType.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Shopsys\FrameworkBundle\Form;
+
+use Shopsys\FrameworkBundle\Form\Locale\LocalizedType;
+
+class LocalizedFullWidthType extends LocalizedType
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/LocalizedFullWidthType.php]: seems that the template for localizedType is too different

File: packages/framework/src/Resources/views/Admin/Form/theme.html.twig [hunk 4]
```diff
@@ -1062,3 +1062,27 @@
         </div>
     </div>
 {% endblock display_only_url_widget %}
+
+
+{% block localized_full_width_label %}{% endblock localized_full_width_label %}
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Resources/views/Admin/Form/theme.html.twig]: yes it was not necessary

File: packages/framework/src/Form/WarningMessageType.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    {
+        $resolver
+            ->setDefaults([
+                'mapped' => false,
+                'required' => false,
+            ]);
+    }
+
+    /**
+     * @return null|string
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/WarningMessageType.php]: fixed

File: packages/framework/src/Form/ProductCalculatedPricesType.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+use Shopsys\FrameworkBundle\Model\Product\Product;
+use Shopsys\FrameworkBundle\Model\Product\ProductFacade;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class ProductCalculatedPricesType extends AbstractType
+{
+    /** @var \Shopsys\FrameworkBundle\Model\Product\Pricing\AdminProductPriceCalculationFacade */
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/ProductCalculatedPricesType.php]: i repaired also the older ones

File: packages/framework/src/Form/DisplayOnlyUrlType.php [hunk 7]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+namespace Shopsys\FrameworkBundle\Form;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class DisplayOnlyUrlType extends AbstractType
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/DisplayOnlyUrlType.php]: seems that DisplayOnlyType got different implementation.

File: packages/framework/src/Form/LocalizedFullWidthType.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/LocalizedFullWidthType.php]: yes i use everything from it except template

File: CHANGELOG.md [hunk 8]
```diff
@@ -45,6 +45,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
     - ProductsType can be used for variants group
     - DisplayOnlyType can be used for disabled groups
     - PluginCrudExtensionFacade was modified to have correct design in administration
+    - templates of FormTypes were modified for use with ProductFormType elements
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply CHANGELOG.md]: extended via more sub rows

File: packages/framework/src/Form/Admin/Product/ProductEditFormType.php [hunk 9]
```diff
@@ -76,12 +85,78 @@ public function __construct(
...
             ])
+                ->addViewTransformer($this->productParameterValueToProductParameterValuesLocalizedTransformer));
+
+        $builder->add($builderParametersGroup);
+
+        $builder->add($seoGroup);
+
+        $pricesGroup = $builder->get('productData')->get('pricesGroup');
+
+        $pricesHolder = $pricesGroup->get('priceHolder');
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/Admin/Product/ProductEditFormType.php]: renamed to productCalculatedPricesGroup

File: packages/framework/src/Form/Admin/Product/ProductEditFormType.php [hunk 10]
```diff
@@ -148,21 +218,23 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                             'groups' => [self::VALIDATION_GROUP_MANUAL_PRICE_CALCULATION],
                         ]),
                     ],
+                    'label' => $pricingGroup->getName(),
                 ]);
         }
 
         if ($editedProduct !== null && $editedProduct->isMainVariant()) {
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply packages/framework/src/Form/Admin/Product/ProductEditFormType.php]: kept

File: CHANGELOG.md [hunk 11]
```diff
@@ -44,6 +44,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - [#375 ProductFormType should be exensible](https://github.com/shopsys/shopsys/pull/375)
     - ProductsType can be used for variants group
     - DisplayOnlyType can be used for disabled groups
+    - PluginCrudExtensionFacade was modified to have correct design in administration
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply CHANGELOG.md]: explained in more details

File: CHANGELOG.md [hunk 12]
```diff
@@ -43,6 +43,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 #### Changed
 - [#375 ProductFormType should be exensible](https://github.com/shopsys/shopsys/pull/375)
     - ProductsType can be used for variants group
+    - DisplayOnlyType can be used for disabled groups
```
[the remark this answers is not in the document]
Comment (boris-brtan) [reply CHANGELOG.md]: added info

File: packages/framework/src/Form/Admin/Product/ProductFormType.php [hunk 13]
```diff
@@ -94,48 +111,235 @@ public function __construct(
...
+        return $builderDescriptionGroup;
+    }
+
+    /**
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     * @param null|Product $product
+     * @param array $disabledItemInMainVariantAttr
+     * @return \Symfony\Component\Form\FormBuilderInterface
+     */
+    private function createDisplayAvaiabilityGroup(FormBuilderInterface $builder, ?Product $product, $disabledItemInMainVariantAttr = [])
```
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply packages/framework/src/Form/Admin/Product/ProductFormType.php]: :+1:

File: packages/framework/src/Form/Admin/Product/ProductFormType.php [hunk 14]
```diff
@@ -94,48 +111,235 @@ public function __construct(
...
+    {
+        $productMainCategoriesIndexedByDomainId = null;
```
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply packages/framework/src/Form/Admin/Product/ProductFormType.php]: :+1:

File: packages/framework/src/Resources/views/Admin/Form/theme.html.twig [hunk 15]
```diff
@@ -66,14 +66,14 @@
...
+            <div class="form-line{{ attr.disabledField is defined ? ' form-input-disabled form-line--disabled' }}">
             {{ form_errors(form) }}
             {{ form_label(form, label) }}
         {% endif %}
         {{ form_widget(form) }}
-        {% if disabledField is defined %}
-            <span class="form-disabled {{ disabledField.class }} js-tooltip" data-toggle="tooltip" data-placement="{{ disabledField.placement|default('right') }}"
-                  title="{{ disabledField.title }}">
+        {% if attr.disabledField is defined %}
+            <span class="form-disabled {{ attr.disabledFieldClass|default('') }} js-tooltip" data-toggle="tooltip" data-placement="{{ attr.disabledFieldPlacement|default('right') }}"
```
Comment (vitek-rostislav) [inline packages/framework/src/Resources/views/Admin/Form/theme.html.twig]: `disabledFieldPlacement` is not defined anywhere, how come?

**Comment by boris-brtan** [comment]
main for domains in https://github.com/shopsys/shopsys/pull/375/commits/02ca4d6d40d4f7fab9eeca71bb81f8b425a26f72

File: packages/framework/src/Resources/views/Admin/Form/theme.html.twig [hunk 16]
```diff
@@ -150,7 +150,14 @@
 
 {% block multidomain_form_rows -%}
     {% for domainId, child in form %}
+        {% if child.vars.attr.show_label is defined %}
+            <div class="form-line__line">
+            {{- form_label(child) -}}
```
Comment (vitek-rostislav) [inline packages/framework/src/Resources/views/Admin/Form/theme.html.twig]: this should be indented :slightly_smiling_face:

File: packages/framework/src/Form/Admin/Product/ProductFormType.php [hunk 17]
```diff
@@ -246,176 +507,247 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+
+        return $builderPricesGroup;
+    }
+
+    /**
+     * @param \Symfony\Component\Form\FormBuilderInterface $builder
+     * @param null|Product $product
+     * @return \Symfony\Component\Form\FormBuilderInterface
+     */
+    private function createSeoFormGroup(FormBuilderInterface $builder, ?Product $product)
```
Comment (vitek-rostislav) [inline packages/framework/src/Form/Admin/Product/ProductFormType.php]: the name is inconsistent with other methods "create*Group" - you should omit the "Form" -> `createSeoGroup()` :slightly_smiling_face:

**Comment by vitek-rostislav** [comment]
Hi @boris-brtan, one more thing - the entry in the `UPGRADE.md` - there should be some upgrading notes if necessary. You just wrote "ProductFormType and ProductEditFormType are now extensible" - it has no added value for me so please either write more instructions what to do when upgrading or remove the entry from the file.

File: packages/framework/src/Resources/views/Admin/Form/theme.html.twig (hunk 15 above)
Comment (boris-brtan) [reply packages/framework/src/Resources/views/Admin/Form/theme.html.twig]: i think it is ready to use feature.

**Comment by boris-brtan** [comment]
Hi @vitek-rostislav,
yes it seems that the information about the change is not needed in changelog.

**Pull request merged**